File: ocaml/supervisor.ml
(* supervisor.ml — OTP-style supervision, itself just a process.
   The supervisor is not a special runtime object: it spawn/link/receive
   like everyone else, traps exits, and restarts according to strategy. *)

open Actor

type strategy = One_for_one | One_for_all | Rest_for_one | Simple_one_for_one
type restart = Permanent | Transient | Temporary

type child_spec = {
  id : string;
  start : unit -> unit;
  restart : restart;
}

let child ?(restart = Permanent) id start = { id; start; restart }

type intensity = { max : int; period : float }

let rec take_recent now period = function
  | t :: rest when now -. t <= period -> t :: take_recent now period rest
  | _ -> []

(* one_for_one: only the dead child is restarted.
   one_for_all: kill the rest, restart everyone.
   Intensity: more than [max] restarts in [period] seconds
   and the supervisor itself dies — the next level takes over. *)

let supervise ?(strategy = One_for_one)
    ?(intensity = { max = 6; period = 12.0 })
    ?(name = "sup") specs =
  register name;
  trap_exit true;
  let spawn_one spec = spawn ~name:spec.id ~link:true spec.start in
  let children =
    specs |> List.map (fun spec -> spec, spawn_one spec, ([] : float list))
  in
  let state = ref children in
  let rec loop () =
    match receive (function Exit _ | Crash -> true | _ -> false) () with
    | Crash -> failwith "supervisor abort"
    | Exit (pid, reason) ->
        let rec restart acc = function
          | [] -> List.rev acc
          | (spec, child_pid, hist) :: rest when child_pid = pid ->
              let skip =
                spec.restart = Temporary
                || (spec.restart = Transient && reason = "normal")
              in
              if skip then restart acc rest
              else
                let t = now () in
                let hist = t :: take_recent t intensity.period hist in
                if List.length hist > intensity.max then
                  failwith "intensity exceeded"
                else begin
                  (match strategy with
                   | One_for_all ->
                       List.iter (fun (_, p, _) -> send p Crash) acc;
                       List.iter (fun (_, p, _) -> send p Crash) rest
                   | _ -> ());
                  let child_pid = spawn_one spec in
                  restart ((spec, child_pid, hist) :: acc) rest
                end
          | row :: rest -> restart (row :: acc) rest
        in
        state := restart [] !state;
        loop ()
    | _ -> loop ()
  in
  loop ()

(* simple_one_for_one: one template, Start_child clones it.
   Children are Temporary — they are not restarted.  Used for
   per-request processes (here: each ticket / order). *)

type msg += Start_child of (unit -> unit) * string

let simple_one_for_one ?(name = "sofs") () =
  register name;
  trap_exit true;
  let rec loop () =
    match receive (function Start_child _ | Exit _ | Crash -> true | _ -> false) () with
    | Crash -> failwith "service abort"
    | Start_child (fn, id) ->
        ignore (spawn ~name:id ~link:true fn);
        loop ()
    | Exit (_, _) ->
        (* temporary: forget the child *)
        loop ()
    | _ -> loop ()
  in
  loop ()

File: src/lib/actor/supervisor.ts
import { fx, type MenuItem, type Msg, type Pid, type ProcFn } from "./types.ts";

export type Strategy = "one_for_one" | "one_for_all" | "rest_for_one";
export type Restart = "permanent" | "transient" | "temporary";

export type ChildSpec = {
  id: string;
  start: ProcFn;
  restart: Restart;
};

type ChildRow = { spec: ChildSpec; pid: Pid; hist: number[] };

export function supervise(opts: {
  name: string;
  specs: ChildSpec[];
  strategy?: Strategy;
  intensity?: number;
  period?: number;
}): ProcFn {
  const strategy = opts.strategy ?? "one_for_one";
  const intensity = opts.intensity ?? 6;
  const period = opts.period ?? 12000;
  return function* () {
    yield fx.register(opts.name, "supervisor.ml:register");
    yield fx.trap_exit(true, "supervisor.ml:trap_exit");
    const children: ChildRow[] = [];
    for (const spec of opts.specs) {
      const pid = (yield fx.spawn(
        spec.id,
        spec.start,
        "supervisor.ml:spawn",
        true,
      )) as Pid;
      children.push({ spec, pid, hist: [] });
    }
    const shutting = new Set<Pid>();
    while (true) {
      const msg = (yield fx.receive(
        (m) => m.t === "EXIT" || m.t === "Crash",
        "supervisor.ml:receive",
      )) as Msg;
      if (msg.t === "Crash") throw new Error("supervisor abort");
      if (msg.t !== "EXIT") continue;
      if (shutting.has(msg.pid)) {
        shutting.delete(msg.pid);
        continue;
      }
      const idx = children.findIndex((c) => c.pid === msg.pid);
      if (idx < 0) continue;
      const row = children[idx]!;
      const skip =
        row.spec.restart === "temporary" ||
        (row.spec.restart === "transient" && msg.reason === "normal");
      if (skip) {
        children.splice(idx, 1);
        continue;
      }
      const t = (yield fx.now("supervisor.ml:intensity")) as number;
      row.hist = [t, ...row.hist.filter((h) => t - h <= period)];
      if (row.hist.length > intensity) throw new Error("intensity exceeded");

      let from = idx;
      let to = idx;
      if (strategy === "one_for_all") {
        from = 0;
        to = children.length - 1;
      } else if (strategy === "rest_for_one") {
        from = idx;
        to = children.length - 1;
      }

      for (let i = from; i <= to; i++) {
        if (i === idx) continue;
        const sib = children[i]!;
        shutting.add(sib.pid);
        yield fx.exit_pid(sib.pid, "shutdown", "supervisor.ml:shutdown");
      }

      for (let i = from; i <= to; i++) {
        const child = children[i]!;
        if (i !== idx && child.spec.restart === "temporary") continue;
        const pid = (yield fx.spawn(
          child.spec.id,
          child.spec.start,
          "supervisor.ml:restart",
          true,
        )) as Pid;
        child.pid = pid;
      }
    }
  };
}

export function simpleOneForOne(
  name: string,
  make: (item: MenuItem, seq: number) => { id: string; fn: ProcFn },
): ProcFn {
  return function* () {
    yield fx.register(name, "supervisor.ml:sofs.register");
    yield fx.trap_exit(true, "supervisor.ml:sofs.trap_exit");
    let n = 1;
    while (true) {
      const msg = (yield fx.receive(
        (m) => m.t === "StartChild" || m.t === "EXIT" || m.t === "Crash",
        "supervisor.ml:sofs.receive",
      )) as Msg;
      if (msg.t === "Crash") throw new Error("service abort");
      if (msg.t === "StartChild") {
        const child = make(msg.item, n++);
        yield fx.spawn(child.id, child.fn, "supervisor.ml:sofs.spawn", true);
      }
    }
  };
}
